selftests/functional/test_statuses.py

```diff
@@ -0,0 +1,291 @@
+                                         'test pre',
+                                         'test post'
+                                         'teardown pre',
+                                         'teardown post']),
+                    'kill-test-27e3': ('ERROR',
+                                       ['setup pre',
+                                        'setup post',
+                                        'test pre'],
+                                       ['test post',
+                                        'teardown pre',
+                                        'teardown post']),
+                    'kill-teardown-94c9': ('ERROR',
+                                           ['setup pre',
+                                            'setup post',
+                                            'test pre',
+                                            'test post',
+                                            'teardown pre'],
+                                           ['teardown post']),
+                    }
+
+
+class TestStatuses(unittest.TestCase):
+
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp(prefix='avocado_' + __name__)
+        test_file = script.TemporaryScript('test.py', TEST_CONTENT)
+        test_file.save()
+        yaml_file = script.TemporaryScript('test.yaml', YAML_CONTENT)
+        yaml_file.save()
+
+        cmd = ('%s run %s -m %s --sysinfo=off --job-results-dir %s --json -' %
+               (AVOCADO, test_file, yaml_file, self.tmpdir))
+
+        results = process.run(cmd, ignore_status=True)
+        self.results = json.loads(results.stdout)
+
+    def test(self):
+        missing_tests = []
+
+        # Testing each individual test results
+        for test in self.results['tests']:
+            variant = test['id'].split(';')[1]
+            expected = EXPECTED_RESULTS.get(variant, False)
+            if not expected:
+                missing_tests.append(variant)
+            else:
+                self._check_test(test, expected)
+
+        # Testing if all variants were covered
+        missing_msg = ' '.join(missing_tests)
+        self.assertEqual(missing_msg, '',
+                         "Expected results not found for variants: %s" %
+                         missing_msg)
+
+    def _check_test(self, test, expected):
+        variant = test['id'].split(';')[1]
+        self.assertEqual(expected[0], test['status'],
+                         "Status error: '%s' != '%s' (%s)" %
+                         (expected[0], test['status'], variant))
+        debug_log = open(test['logfile'], 'r').read()
+        for msg in expected[1]:
+            self.assertIn(msg, debug_log,
+                          "Message '%s' should be in the log (%s)" %
+                          (msg, variant))
+        for msg in expected[2]:
+            self.assertNotIn(msg, debug_log,
+                             "Message '%s' should not be in the log (%s)" %
+                             (msg, variant))
```

Comment:
would you please provide all necessary details in case of failure? I'd expect the json output and job.log to be able to tell what was actually going on.